src/Core/Configurations/RuntimeConfigValidator.cs

```diff
@@ -830,11 +830,161 @@ public void ValidateRelationshipsInConfig(RuntimeConfig runtimeConfig, IMetadata
+                                $"but target fields that are null.",
+                            statusCode: HttpStatusCode.ServiceUnavailable,
+                            subStatusCode: DataApiBuilderException.SubStatusCodes.ConfigValidationError));
+                    }
+
+                    // When both source and linkingSource fields exist for a many to many relationship their size must match.
+                    if (relationship.SourceFields is not null && relationship.LinkingSourceFields is not null)
+                    {
+                        if (relationship.SourceFields.Length != relationship.LinkingSourceFields.Length)
+                        {
+                            // Validation to ensure that if source and linking source fields exist, that they have the same number of fields. 
+                            HandleOrRecordException(new DataApiBuilderException(
+                                message: $"Entity: {entityName} has a many to many relationship: {relationshipName} with {relationship.SourceFields.Length} " +
+                                    $"source fields defined, but {relationship.LinkingSourceFields.Length} linking source fields defined.",
+                                statusCode: HttpStatusCode.ServiceUnavailable,
+                                subStatusCode: DataApiBuilderException.SubStatusCodes.ConfigValidationError));
+                        }
+                    }
+
+                    // When both target and linkingTarget fields exist for a many to many relationship their size must match.
+                    if (relationship.TargetFields is not null && relationship.LinkingTargetFields is not null)
+                    {
+                        if (relationship.TargetFields.Length != relationship.LinkingTargetFields.Length)
+                        {
+                            // Validation to ensure that if source and linking source fields exist, that they have the same number of fields. 
+                            HandleOrRecordException(new DataApiBuilderException(
+                                message: $"Entity: {entityName} has a many to many relationship: {relationshipName} with {relationship.TargetFields.Length} " +
+                                    $"target fields defined, but {relationship.LinkingTargetFields.Length} linking target fields defined.",
+                                statusCode: HttpStatusCode.ServiceUnavailable,
+                                subStatusCode: DataApiBuilderException.SubStatusCodes.ConfigValidationError));
+                        }
+                    }
+                }
+
                 // Validation to ensure DatabaseObject is correctly inferred from the entity name.
                 DatabaseObject? sourceObject, targetObject;
                 if (!sqlMetadataProvider.EntityToDatabaseObject.TryGetValue(entityName, out sourceObject))
```

Comment:
line 1046, where we are just doing _logger.Debug, can we move this out of this method? this method is too big now.

Reply:
I am not sure what you mean, this seems to log information about the relationship in non-hosted scenarios, do you want to remove the logging entirely?

![image](https://github.com/Azure/data-api-builder/assets/93220300/60b62ab5-a0ab-41e0-9cc6-2e3125ea1773)

Reply:
I am not sure what you mean by moving this out of this method, the information that is being logged here is taken from the calls made just previous to the logging. Where would it go if not here?

![image](https://github.com/Azure/data-api-builder/assets/93220300/8bd63fb9-9d48-4806-930d-67d9314fbac7)

Reply:
if there is cleanup needed outside these changes we can add to backlog.